src/components/tl/ucp/allgather/allgather_sparbit.c

```diff
@@ -0,0 +1,156 @@
+    ignore_steps = (~((uint32_t)tsize >> last_ignore) | 1) << last_ignore;
+
+    while (i < tsize_log) {
+        data_expected = task->allgather_sparbit.data_expected;
+
+        distance = 1 << (tsize_log - 1);
+        distance >>= i; // restore distance in case of continuation depending on step
+
+        recvfrom  = (trank + tsize - distance) % tsize;
+        sendto    = (trank + distance) % tsize;
+        exclusion = (distance & ignore_steps) == distance;
+
+        for (transfer_count = 0; transfer_count < data_expected - exclusion;
+             transfer_count++) {
+            tmprecv = PTR_OFFSET(
+                rbuf, (trank - (2 * transfer_count + 1) * distance + tsize) %
+                          tsize * data_size);
+            tmpsend = PTR_OFFSET(
+                rbuf, (trank - 2 * transfer_count * distance + tsize) % tsize *
+                          data_size);
+            /* Sendreceive */
+            UCPCHECK_GOTO(ucc_tl_ucp_send_nb(tmpsend, data_size, rmem, sendto,
+                                             team, task),
+                          task, out);
+            UCPCHECK_GOTO(ucc_tl_ucp_recv_nb(tmprecv, data_size, rmem, recvfrom,
+                                             team, task),
+                          task, out);
+        }
+
+        task->allgather_sparbit.data_expected =
+            (data_expected << 1) - exclusion;
+        task->allgather_sparbit.i++;
+        // check if we could make one more step right now or we should yeld task
+        if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
+            return;
+        }
+        i = task->allgather_sparbit.i;
+    }
+
+    ucc_assert(UCC_TL_UCP_TASK_P2P_COMPLETE(task));
+    task->super.status = UCC_OK;
+
+out:
+    UCC_TL_UCP_PROFILE_REQUEST_EVENT(coll_task, "ucp_allgather_sparbit_done",
+                                     0);
+}
+
+ucc_status_t ucc_tl_ucp_allgather_sparbit_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task      = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team      = TASK_TEAM(task);
+    size_t             count     = TASK_ARGS(task).dst.info.count;
+    void              *sbuf      = TASK_ARGS(task).src.info.buffer;
+    void              *rbuf      = TASK_ARGS(task).dst.info.buffer;
+    ucc_memory_type_t  smem      = TASK_ARGS(task).src.info.mem_type;
+    ucc_memory_type_t  rmem      = TASK_ARGS(task).dst.info.mem_type;
+    ucc_datatype_t     dt        = TASK_ARGS(task).dst.info.datatype;
+    ucc_rank_t         trank     = UCC_TL_TEAM_RANK(team);
+    ucc_rank_t         tsize     = UCC_TL_TEAM_SIZE(team);
+    size_t             data_size = (count / tsize) * ucc_dt_size(dt);
+    ucc_status_t       status;
+
+    UCC_TL_UCP_PROFILE_REQUEST_EVENT(coll_task, "ucp_allgather_sparbit_start",
+                                     0);
+    ucc_tl_ucp_task_reset(task, UCC_INPROGRESS);
+
+    /* initial step: copy data on non root ranks to the beginning of buffer */
```

Comment:
This comment doesn't seem to describe what the code does. 
1. I don't see a check root/non-root
2. it doesn't seem to be the beginning of the buffer, rather a copy from src to dst buffer at the rank offset
3. what if it is inplace?

Reply:
will remove, because it is copypaste from bruck algo